lib/internal/stream_base_commons.js

```diff
@@ -134,29 +135,28 @@ function writevGeneric(self, data, cb) {
   // Retain chunks
   if (err === 0) req._chunks = chunks;
 
-  afterWriteDispatched(self, req, err, cb);
+  afterWriteDispatched(self, req, err);
   return req;
 }
 
 function writeGeneric(self, data, encoding, cb) {
-  const req = createWriteWrap(self[kHandle]);
+  const req = createWriteWrap(self[kHandle], cb);
```

Comment:
this kind of tiny subtle change in cb behaviour seems to be a big deal, worth pulling out standalone. cc:@ronag

Reply:
It's a little unclear why this change is needed? Also, it seems this hasn't been rebased against master in a while since there has been [changes](https://github.com/nodejs/node/commit/8f86986985b161b7fb5aa41283d792417329bc2c#diff-704db381d3305bdcd2c6b8b3468acfb7).

Reply:
This bit can definitely be revisited. In fact, eventually I'd like to refactor the entire interface here. Here's the background...

To make writes as efficient as possible, when writes happen on the `QuicStream` duplex, those are passed down to the underlying C++ `QuicStream` `StreamBase` specialization. That is passed as one or more uv_buf_t instances and a WriteReq instance. Those are stored in a linked link without copying the data. That linked list maintains three distinct pointers: 1. The tail (insertion point), 2. The current read position, and 3. The current acknowledge position. As the underlying QUIC implementation serializes data into QUIC packets, the second pointer is advanced. As the endpoint receives acknowledgements from the peer, the third pointer is advanced. As the acknowledge pointer advances, a check is made to see if all of the uv_buf_t's associated with a given WriteReq have been acknowledged. If they have, the WriteReq is completed and the associated callback is invoked. *Theoretically* on the C++ side, the linked list of buffers/WriteReqs can be any length. *In practice*, however, the JavaScript side and the current implementation of `StreamBase` only allows one WriteReq at a time passed to the C++ side, and `StreamBase` does not allow the `QUICStream` to take ownership over the `Buffer` instances that are passed (all we get are `uv_buf_t` views over the Buffer's data). This artificially and unnecessarily limits our throughput. The current implementation exists largely because backpressure is implemented at the JavaScript level, whereas with QUIC it is by far more efficient to rely on the QUIC-level flow control mechanisms and per-Session memory limits.

What I want to move to is a model that, based on a constructor option, the Duplex will eliminate buffering on the JavaScript side and will pass writes directly and immediately on to the C++ side -- which will provide the necessary backpressure to avoid over-buffering. I further want to allow the underlying `QuicBuffer` to take ownership over the `Buffer` instances passed to it so that those are decoupled from the lifetime of the `WriteReq`, allowing us to invoke the callbacks when the written data is serialized and flushed to the network rather than when we receive acknowledgement for it.

Reply:
> _Theoretically_ on the C++ side, the linked list of buffers/WriteReqs can be any length.

I think we’ve moved away from that being possible, fwiw.

Reply:
Which being possible? The current implementation of `QuicBuffer` supports the model I'd like to get to but we're limited by the current implementation of `StreamBase` and `Writable`.